feat: O3DE test benchmark

jmatejcz opened the pull request:
## Purpose
Add Test Benchmark for O3DE simulation
Add framework for easier benchmark creation

## Proposed Changes
Majority of file changes are only formatting changes, main changes are in src/rai_bench package and src/rai_sim 

This PR contains new package rai_bench. It introduces classes like  Task, Scenario and Benchmark(src/rai_bench/rai_bench/benchmark_model.py), which are frame for creating benchmarks and it also contains O3DE Test Benchmark which is a test implementation of a benchmark for O3DE engine simulation.

O3DE Test Benchmark (`src/rai_bench/rai_bench/o3de_test_bench/`), contains 2 Tasks(`tasks/`) - GrabCarrotTask and PlaceCubesTask  (these tasks implement calculating scores) and 4 scenes(`configs/`) for O3DE robotic arm simulation. 

Example of how can scenarios and benchmark be defined and run  can be found in `rai_bench/rai_bench/main.py`

Additional changes:
- Subclass of Bridge for Arm Manipulation-> O3DEngineArmManipulationBridge
- validation if required services are available -> _is_robotic_stack_ready() function in SimulationBridge
- fix a bug when get_transform returned same values every time-> https://github.com/RobotecAI/rai/issues/430

## Issues

## Testing
the binary to simulation can be downloaded from here: [humble](https://robotecai.sharepoint.com/Shared%20Documents/Forms/AllItems.aspx?id=%2FShared%20Documents%2FRAIManipulationDemoBenchmark%5Fjammyhumble%20%28old%29%2Ezip&parent=%2FShared%20Documents&p=true&ga=1)


To run this benchmark you need o3de bridge config file in `src/rai_bench/rai_bench/o3de_test_bench/configs/o3de_config.yaml`, this config file should include fields:
```yaml
binary_path: ...
robotic_stack_command: ros2 launch examples/manipulation-demo-no-binary.launch.py
required_services:
  - /grounding_dino_classify
  - /grounded_sam_segment
  - /manipulator_move_to
  - /spawn_entity
  - /delete_entity
required_topics:
  - /color_image5
  - /depth_image5
  - /color_camera_info5
required_actions: []
```
to install and run:
```
poetry install --with openset
colcon build --symlink-install
source setup_shell.sh
python src/rai_bench/rai_bench/main.py
```
Logs about the progess of benchmark can be found in 'src/rai_bench/o3de_test_bench/benchmark_agent.log'

You should be able to see: 
- properly started simulation
- 4 defined scenarios
- logs about the actions taken by agent and scores achieved in each scenario
After finishing all scenarios, simulation should close and all nodes and services should also shutdown after no more than 30 seconds

MagdalenaKotynia committed:
feat: benchmark class models

Co-authored-by: Jakub Matejczyk <jakub.matejczyk@robotec.ai>

chore: naming changes

Co-authored-by: Jakub Matejczyk <jakub.matejczyk@robotec.ai>

chore: naming changes

Co-authored-by: Jakub Matejczyk <jakub.matejczyk@robotec.ai>

chore: redefine benchmark model

Co-authored-by: Jakub Matejczyk <jakub.matejczyk@robotec.ai>
Co-authored-by: Kacper Dąbrowski <kacper.dabrowski@robotec.ai>

refactor: SceneConfig to BaseModel class

feat: add SceneSetup to store initial scene setup

build: poetry initialization of rai_benchmarks and rai_simulations

chore: add licence lines

build: create packages from rai_benchmarks and rai_simulations

chore: removed mistakenly added file

feat: scene config implementation

Add O3DEEngineConnector features

Signed-off-by: Kacper Dąbrowski <kacper.dabrowski@robotec.ai>

Remove an unused file

Signed-off-by: Kacper Dąbrowski <kacper.dabrowski@robotec.ai>

Add binary path caching

Signed-off-by: Kacper Dąbrowski <kacper.dabrowski@robotec.ai>

Add two example scenes

Signed-off-by: Kacper Dąbrowski <kacper.dabrowski@robotec.ai>

feat: replace binary path with ros2 launch command + binary path

refactor: renamed rai_sim and rai_bench

fix: fixed shutdown of binary

chore: make pose mandatory

chore: remove rai_bench because it is developed on another branch

ci: add missing license lines

laoding and spawning frm benchmark

rabsed

naming change

grab xyz benchmark

benchmarks run new sim every scenario for now

MagdalenaKotynia committed:
feat: benchmark class models

Co-authored-by: Jakub Matejczyk <jakub.matejczyk@robotec.ai>

chore: naming changes

Co-authored-by: Jakub Matejczyk <jakub.matejczyk@robotec.ai>

chore: naming changes

Co-authored-by: Jakub Matejczyk <jakub.matejczyk@robotec.ai>

chore: redefine benchmark model

Co-authored-by: Jakub Matejczyk <jakub.matejczyk@robotec.ai>
Co-authored-by: Kacper Dąbrowski <kacper.dabrowski@robotec.ai>

refactor: SceneConfig to BaseModel class

feat: add SceneSetup to store initial scene setup

build: poetry initialization of rai_benchmarks and rai_simulations

chore: add licence lines

build: create packages from rai_benchmarks and rai_simulations

chore: removed mistakenly added file

feat: scene config implementation

Add O3DEEngineConnector features

Signed-off-by: Kacper Dąbrowski <kacper.dabrowski@robotec.ai>

Remove an unused file

Signed-off-by: Kacper Dąbrowski <kacper.dabrowski@robotec.ai>

Add binary path caching

Signed-off-by: Kacper Dąbrowski <kacper.dabrowski@robotec.ai>

Add two example scenes

Signed-off-by: Kacper Dąbrowski <kacper.dabrowski@robotec.ai>

feat: replace binary path with ros2 launch command + binary path

refactor: renamed rai_sim and rai_bench

fix: fixed shutdown of binary

chore: make pose mandatory

chore: remove rai_bench because it is developed on another branch

ci: add missing license lines

laoding and spawning frm benchmark

rabsed

naming change

grab xyz benchmark

benchmarks run new sim every scenario for now

maciejmajek committed:
refactor(partial): refactor openset tools to rai2.0 ros2ariconnector api

maciejmajek committed:
refactor: finalize with simplifications

maciejmajek committed:
fix: missing argument, missing client

jmatejcz committed:
adjust to new tools

boczekbartek committed:
workaround: don't remove subscription due to rclpy issue

https://github.com/ros2/rclpy/issues/1142

boczekbartek committed:
replace ros2 future waiting with more robust callback based mechanism

boczekbartek committed:
remove waiting from launch

boczekbartek committed:
fix order in connector shutdown

```
Exception in thread Thread-1 (spin):
Traceback (most recent call last):
  File "/usr/lib/python3.10/threading.py", line 1016, in _bootstrap_inner
    self.run()
  File "/usr/lib/python3.10/threading.py", line 953, in run
    self._target(*self._args, **self._kwargs)
  File "/opt/ros/humble/local/lib/python3.10/dist-packages/rclpy/executors.py", line 294, in spin
    self.spin_once()
  File "/opt/ros/humble/local/lib/python3.10/dist-packages/rclpy/executors.py", line 794, in spin_once
    self._spin_once_impl(timeout_sec)
  File "/opt/ros/humble/local/lib/python3.10/dist-packages/rclpy/executors.py", line 786, in _spin_once_impl
    self._executor.submit(handler)
  File "/usr/lib/python3.10/concurrent/futures/thread.py", line 167, in submit
    raise RuntimeError('cannot schedule new futures after shutdown')
RuntimeError: cannot schedule new futures after shutdown
```

boczekbartek committed:
refactor: cleanup manipulation launchfile

boczekbartek committed:
chore: pre-commit

boczekbartek committed:
Remove psutil conflicting __del__ method from o3de connector

See this issue: https://github.com/giampaolo/psutil/issues/2437

boczekbartek committed:
stream langgraph agent

boczekbartek committed:
extent output of tool runner

boczekbartek committed:
enable all scenarios in the benchmark

jmatejcz committed:
feat: added client to arm manipulator to move arm between scenarios
removed name in launch_robotic_manipulation to avoid misnaming nodes

jmatejcz committed:
feat: add GrabCarrot and RedCubes Tasks
add main to run bechmark
add helper functions to Task

jmatejcz committed:
refactor: getting initial and current objects positions separatly

jmatejcz committed:
style: added license

jmatejcz committed:
chore: delete old code

jmatejcz committed:
reafactor: refactored red cube metric

jmatejcz committed:
refactor: refactored carrot metric

jmatejcz committed:
feat: add entities mismatch exception

jmatejcz committed:
feat: add logger to task

jmatejcz committed:
feat: add counting time of scenario execution

rachwalk commented on imgui.ini:
```diff
@@ -0,0 +1,5 @@
+Pos=60,60
+Size=400,400
+Collapsed=0
+
```
Why is this file included?

jmatejcz replied:
this file is generated by simulation i guees, i can add it to gitignore

rachwalk commented on src/rai_bench/README.md:
Why is an empty README included

jmatejcz replied:
filled readme

rachwalk commented on src/rai_bench/rai_bench/benchmark_model.py:
```diff
@@ -0,0 +1,195 @@
+)
+
+
+SimulationConnectorT = TypeVar("SimulationConnectorT", bound=SimulationBridge)
```
Rename this to avoid confusion with rai BaseConnector derived types

rachwalk commented on src/rai_bench/rai_bench/benchmark_model.py:
```diff
@@ -0,0 +1,195 @@
+                # TODO (jm) figure out how to get number of tool calls
+            te = time.perf_counter()
+
+            self.engine_connector.get_scene_state()
```
What is this for?

rachwalk commented on src/rai_bench/rai_bench/benchmark_model.py:
```diff
@@ -0,0 +1,195 @@
+    scene_config: SimulationConfig
+    model_config = ConfigDict(
+        arbitrary_types_allowed=True
+    )  # pydantic does not support ABC classes
```
As far as I can see this parameter is never used, what should it be used for?

jmatejcz replied:
model_config field is used to define additional configurations to a model. In this case the param arbitrary_types_allowed=True, allows a field with an abstract class like SimulationConfig. Without it pydantic will throw validation error

rachwalk replied:
Please mark this field clearly with a `# TODO:` explaining functioanality this field is intended for.

jmatejcz replied:
done

rachwalk commented on src/rai_bench/o3de_test_bench/main.py:
```diff
@@ -0,0 +1,302 @@
+from pathlib import Path
+
+
+class GrabCarrotTask(Task):
```
It would be a good idea to create a module with Tasks so they can be imported into different benchmark runtimes.

jmatejcz replied:
Do you mean to create a separate folder still inside o3de_test_bench? or to define specific tasks in rai_bench package?

rachwalk replied:
Yes I mean create a separate folder with __init__.py (i.e. a submodule) inside the `o3de_test_bench`

jmatejcz replied:
i moved tasks to separate folder, also moved whole o3de_test_branch to rai_bench package so the benchamark/ tasks can be imported across project. main.py to setup and run benchmark moved to src/rai_bench/main.py, so still outside package

rachwalk commented on src/rai_bench/o3de_test_bench/scene1.yaml:
It would be a good idea to move these to a separate folder to keep things organised.

rachwalk commented on src/rai_bench/o3de_test_bench/o3de_config.yaml:
Why is this file included?

jmatejcz replied:
it is needed to simulation configuration

rachwalk commented on src/rai_bench/o3de_test_bench/main.py:
```diff
@@ -0,0 +1,302 @@
+            task=RedCubesTask(logger=bench_logger),
+            scene_config=multiple_cubes_scene_config,
+        ),
+    ]
```
It would be a good idea to have these load dynamically to allow for easy configuration of scenes/tasks.

jmatejcz replied:
do you mean loading from cmd arguments, from config file, or what?

rachwalk replied:
I mean a functionality to auto create scenarios given a list of Tasks and a list of scene paths

jmatejcz replied:
done

rachwalk reviewed (changes requested):
Overall good job, I have left a few comments regarding code structure, and I am unsure if all the committed files should be included.

MagdalenaKotynia commented on src/rai_bench/rai_bench/benchmark_model.py:
```diff
@@ -0,0 +1,195 @@
+        super().__init__(message)
+
+
+class Task(ABC, Generic[SimulationConnectorT]):
```
Task class should not be bound with any specific SimulationBridge, it should be reused with different implementations of SimulationBridge. Please apply appropriate typing for this class and the calculate_result method.

jmatejcz replied:
Okay so i will modify the Task to take SimulationConfig as argument, and from this i will extract initial_positions of objects.

jmatejcz replied:
Second thing is - as spawned_entities are unique to O3De , methods of SimulaitonBride should return Entities, not spawnedEntities

MagdalenaKotynia replied:
I think that Task does not need to take SimulationConfig as argument - Scenario uses SimulationConfig and bounds it with Task (btw please rename scene_config to simulation_config according to the changes in naming convention). Task itself should be SimulationConfig-agnostic, current interface is ok, just typing is inappropriate. 

Second, yes, good point. I noticed it yesterday when reviewing your PR and I decided to move spawned_entities to base SimulationBridge class because it is needed for all simulations if we want to compare them. Please rebase to these changes.

jmatejcz replied:
i refactored it to match your new code, task now uses spawned_entities attribute from Bridges

MagdalenaKotynia commented on src/rai_bench/rai_bench/benchmark_model.py:
```diff
@@ -0,0 +1,195 @@
+        if logger:
+            self._logger = logger
+        else:
+            self._logger = logging.getLogger(__name__)
```
I think that class Benchmark should be redefined. Currently, it allows for benchmarking scenarios and agents but it does not allow for benchmarking simulations using different implementations of SimulationBridge interface (see requirements in https://github.com/RobotecAI/rai/issues/405#issue-2835184006). What about creating class Benchmark as an interface for implementing specific benchmarks, e.g. SimBenchmark to benchmark different simulations with locked llm model and agent structure; or AgentBenchmark to benchmark different Agents with locked simulation, llm model; etc?

jmatejcz replied:
why it doesn't allow to use different engines? You can assign any Bridge you want

jmatejcz replied:
i will add it as arg to constructor, but the logic stays the same

MagdalenaKotynia replied:
Yes, I  can use any Bridge I want but I cannot benchmark bridges - I cannot pass more than one bridge and test them with locked other parameters.

jmatejcz replied:
oh, okay to maybe accepting list of engines, and running all scenarios for every one of them?

rachwalk replied:
At this point there is no need for such functionality, as long as the benchmark is defined as list[scene] x list[task] x model_config, this can be separately replicated against different engines in separate runs. I would see this functionality being a part of benchmark as overkill.

MagdalenaKotynia commented on src/rai_bench/rai_bench/benchmark_model.py:
```diff
@@ -0,0 +1,195 @@
+        logger: loggers_type | None = None,
+    ) -> None:
+        self.engine_connector: SimulationBridge
+        self.tasks: list[Task] = []
```
For what is self.tasks? It is not used, tasks are already included in the scenarios.

jmatejcz replied:
you are right, removing...

MagdalenaKotynia commented on src/rai_bench/rai_bench/benchmark_model.py:
```diff
@@ -0,0 +1,195 @@
+            te = time.perf_counter()
+
+            self.engine_connector.get_scene_state()
+            result = scenario.task.calculate_result(self.engine_connector)
```
Please append result to the list of results. I would also consider a different structure for storing results, to also store information about the experiment to which a given result belongs.

jmatejcz replied:
done

MagdalenaKotynia commented on src/rai_bench/rai_bench/benchmark_model.py:
```diff
@@ -0,0 +1,195 @@
+        else:
+            self._logger = logging.getLogger(__name__)
+
+    def run_next(self, agent):
```
I think it would be useful to calculate the result also at the beginning of each scenario, to be able to recognize if a good result at the end is not a coincidental effect of a favorable initial arrangement of elements on the scene.
I see that it is somehow handled in the implementations of calculate_result for each task, but I think calculating results with the same metric at the beginning and at the end of the scenario would be a simpler approach - the user wouldn't need to think about it when writing metrics.

jmatejcz replied:
I get your point, you pobably refering to the special case where there is nothing to move and cases where the inital setup is almost proper. But not moving objects that are already in a good place is also a good behaviour, thats why agent can achive 0.5/1 without moving anything, because it is better than displacing objects. For me, the main souce of calculation metrics are these: corrected_objects, misplaced_objects, unchanged_correct, displaced_objects. The score can be than calculated in different manners, for example punishing more displacing objects, but it is specific implementation. So to summarize i know it can be confusing that score is 1.0 when agent done nothing, but it is specific to scene and implementation of task. If user wants to calculate it like that, so be it. I can add the calculation at the start, but it seems weird to me

MagdalenaKotynia replied:
I think that if we calculate the result at the beginning and at the end and compare them, we can also see whether the result has been worsened compared to the initial state. 

Apart from that, current implementation does not take into account all quite probable situations, e.g. 
Initial state of cubes is that two of them are adjacent at one edge of the table, and two of them are adjacent on the second edge of the table. 
**Scenario 1**: Agent does nothing. The score will be **1**, because every cube has one adjacent.
**Scenario 2**: Arm moves one cube from one edge to another and as result we have 3 adjacent cubes on one edge and one cube without any adjacent. The result will be **0.75**. The metric will classify the moved cube as **unchanged_correct** and the cube left without any adjacent as **displaced_objects** what is not true. 

In Scenario 2 agent took some action to make situation better but the result score is worse than that in Scenario 1. Therefore I would suggest to calculate the result more generally with some methods checking clusterization of the objects.
This is also why I see advantages of having the result from the beginning and from the end of the scenario. 

Uh, hard task ;)

jmatejcz replied:
okay, maybe i will define task more clearly, for example - "all cubes have to be adjacent at least to one other cube" or "place all cubes in one place/cluster"

jmatejcz replied:
also the naming  be confusing ex. displaced objects, it is counted as ->  "when the object which was in the correct place at the start, is in a incorrect place at the end", but the naming suggest an action was taken to move it

MagdalenaKotynia replied:
> okay, maybe i will define task more clearly, for example - "all cubes have to be adjacent at least to one other cube" or "place all cubes in one place/cluster"

It is not a problem in task definition but in metric calculation. A more precise prompt will not resolve the problem that I described above (Maybe the agent will behave better but we cannot measure it properly with the current metric. 





> also the naming be confusing ex. displaced objects, it is counted as -> "when the object which was in the correct place at the start, is in a incorrect place at the end", but the naming suggest an action was taken to move it

Actually, the `correct place` phase is still confusing because there is no one correct place for the cubes in this task. The success of this task depends on the relation between the cubes' placement.

jmatejcz replied:
as we talked in the office, the task if defined differently now so this issue is no longer present

MagdalenaKotynia commented on src/rai_bench/o3de_test_bench/main.py:
```diff
@@ -0,0 +1,302 @@
+            return (corrected_objects + unchanged_correct) / num_of_objects
+
+
+def request_to_base_position() -> ManipulatorMoveTo.Request:
```
What about moving this function to O3DEngineArmManipulationBridge and parametrizing it with the initial arm pose?

jmatejcz replied:
i considered this, but there would be a lot of args, but yes it can be done

MagdalenaKotynia replied:
I think that you don't need to move this function as it is but rather refactor this solution somehow to store initial position inside the O3DEngineArmManipulationBridge and reuse it with move_arm when needed.

jmatejcz replied:
okay, done

MagdalenaKotynia commented on src/rai_bench/o3de_test_bench/main.py:
```diff
@@ -0,0 +1,302 @@
+            return (corrected_objects + unchanged_correct) / num_of_objects
+
+
+class RedCubesTask(Task):
```
Rename the class accordingly to the prompt of the task.

MagdalenaKotynia commented on src/rai_bench/o3de_test_bench/main.py:
```diff
@@ -0,0 +1,302 @@
+                "Number of initially spawned entities does not match number of entities present at the end."
+            )
+
+        for ini_carrot in initial_carrots:
```
This check is strictly matching our simulation where we know how the table is located with respect to the origin. Maybe it's overkill for now but I would consider parametrizing it somehow. Maybe the calculate_result method should take some kwargs to be able to reuse the task for simulations with different parameters key for computing result (e.g. in this case the info about coordinates and size of the table top).

jmatejcz replied:
I don't know how the cooridinates work in other simulation engines, so it's difficult to make it work. I think for now leave it as this is, maybe in future when we will work with different engines, I will take this into cosideration

MagdalenaKotynia replied:
I didn't mean other simulation engines but e.g. another o3de robotic arm simulation where the table is located in another way with respect to origin than the current simulation we use, e.g. (x,y) = 0 is not on the  middle of the table as it is assumed in this example. I think we can leave it for now, general differentiating where is left and right is quite hard problem IMO, please just leave a NOTE with info that it is example for particular demo.

jmatejcz replied:
okay

MagdalenaKotynia commented on src/rai_bench/o3de_test_bench/main.py:
```diff
@@ -0,0 +1,302 @@
+
+        if num_of_objects == 0:
+            self.logger.info("No objects to manipulate, returning 1.0")
+            return 1.0
```
I think we should not return 1.0 when there are no objects to manipulate. Maybe None would be better for such situations.

I would also suggest moving this check before for loop - I know that it does not have any effect but IMO it will be more logical to check this condition before going further.

jmatejcz replied:
checking before the loop is a valid point. About the score, i think this should always return score. The score would be 1.0 in this situation as there is nothing to be done, but the agent still can be evaluated by number of tool calls and total time

MagdalenaKotynia commented on src/rai_bench/o3de_test_bench/main.py:
```diff
@@ -0,0 +1,302 @@
+        final_carrots = self.filter_entities_by_prefab_type(
+            scene_state.entities, prefab_types=["carrot"]
+        )
+        num_of_objects = len(initial_carrots)
```
I suggest renaming to num_initial_carrots to be clearer

MagdalenaKotynia reviewed (changes requested):
Thank you for this PR. I have left some comments. 
Apart from that:

* Please apply new naming convention for simulation connectors (see https://github.com/RobotecAI/rai/pull/415, SimulationConnector was renamed to SimulationBridge to avoid confusion with the BaseConnector and its subclasses.

* Please fix typing where possible

Ideas for the future improvement:
* I think it would be very beneficial to save output from camera (from the whole scenario or at least beginning and end of the scenario) to be able to verify metrics with human assessment. It also could be useful for presenting and reporting the results.

* When the final implementation of benchmark interface will be established, the next step could be to design and implement some structure for saving and storing the results (results + info about what was benchmarked + camera images + metadata).

jmatejcz commented:
> Overall good job, I have left a few comments regarding code structure, and I am unsure if all the committed files should be included.

the additiional files propably come from rebases, but you are right, they are not part of this PR, so  I will remove them

jmatejcz committed:
chore: add readme, delete imgui file

jmatejcz committed:
feat: moved configs to separate folder
task takes simulation config as arg now
adjusted typing

jmatejcz commented:
add new occuring issue to the main comment of this PR

jmatejcz committed:
feat: benchmark takes bridge as arg
 adjusted paths to configs
 add move arm request to O3DEngineArmManipulationBridge
adjusted main to task naming change

jmatejcz committed:
chore: ignore o3de config

jmatejcz committed:
style: add notes to task and todos to scenario

jmatejcz committed:
refactor: adjust to naming of Bridges changes

jmatejcz committed:
fix: adjust args passing in calculate_results

jmatejcz committed:
style: moved tasks to separate folder, moved o3de benchmark to rai_bench package

jmatejcz committed:
feat: add validation if scene is suitable for task

jmatejcz committed:
feat: add way to automatically create scenarios

jmatejcz committed:
feat: variables in score calculations rename to better suit purpouse
task defined more clearly

jmatejcz committed:
feat: add storing results

jmatejcz committed:
chore: add license

jmatejcz commented:
found new issue with getting object position from simulation, part of it was due to changes introduced in feat/benchmarking , resolved here: https://github.com/RobotecAI/rai/pull/415/commits/702bb00bd08a3c61364b46a35439ddab7200781c

Now the transform is good, but the position of object is always same as starting position, even when object was moved on simulation

jmatejcz committed:
partially fixed getting position bug

jmatejcz committed:
fix: move tf_listener to constructor in ROS2ARIConnector
This fixes bug when get_transform returns always the same, first transform
Add logs to track retrieved positions

jmatejcz committed:
fix: fix calculations with adjacent objects in Tasks

jmatejcz commented:
> found new issue with getting object position from simulation, part of it was due to changes introduced in feat/benchmarking , resolved here: [702bb00](https://github.com/RobotecAI/rai/commit/702bb00bd08a3c61364b46a35439ddab7200781c)
> 
> Now the transform is good, but the position of object is always same as starting position, even when object was moved on simulation

fixed here: https://github.com/RobotecAI/rai/commit/871f707347b171b4725a5c3daac26ef4b441e7cb

jmatejcz committed:
feat: add counting number of tool calls

rachwalk reviewed (changes requested):
When attempting to run the according to PR description I get an error:

```
Traceback (most recent call last):
  File "/home/krachwal/projects/internal/rai/src/rai_bench/rai_bench/main.py", line 21, in <module>
    from rai_bench.benchmark_model import (
ModuleNotFoundError: No module named 'rai_bench'
```

It seems that poetry install doesn't install the new package.

jmatejcz committed:
chore: delete o3de config, remove from gitignore

jmatejcz committed:
fix: update pyproject toml to include rai bench

jmatejcz commented:
> When attempting to run the according to PR description I get an error:
> 
> ```
> Traceback (most recent call last):
>   File "/home/krachwal/projects/internal/rai/src/rai_bench/rai_bench/main.py", line 21, in <module>
>     from rai_bench.benchmark_model import (
> ModuleNotFoundError: No module named 'rai_bench'
> ```
> 
> It seems that poetry install doesn't install the new package.

added package to pyproject.toml

jmatejcz commented:
the scenario should not run until all services are up ( case when grounding dino is downloading weights)

jmatejcz committed:
formatting changes

MagdalenaKotynia commented on src/rai_bench/rai_bench/benchmark_model.py:
```diff
@@ -0,0 +1,246 @@
+                    print(
+                        f"Could not create Scenario from task: {task.get_prompt()} and simulation_config: {sim_conf}, {e}"
+                    )
+        return scenarios
```
What is the purpose of this method? What does this try-except block handle?

jmatejcz replied:
scenario is validating if task can be paired with scene, if not ValueError is raised

MagdalenaKotynia commented on src/rai_sim/rai_sim/o3de/o3de_bridge.py:
```diff
@@ -304,3 +308,57 @@ def from_ros2_pose(self, pose: Pose) -> PoseModel:
+            request.target_pose.pose.orientation.y = pose.rotation.y
+            request.target_pose.pose.orientation.z = pose.rotation.z
+            request.target_pose.pose.orientation.w = pose.rotation.w
+        else:
```
For what is this default rotation of 180 degrees in the x-axis?

jmatejcz replied:
deleted default rotation, added this rotation to main.py

MagdalenaKotynia commented on src/rai_bench/rai_bench/main.py:
```diff
@@ -0,0 +1,174 @@
+    )
+
+    # custom request to arm
+    base_arm_pose = PoseModel(translation=Translation(x=0.5, y=0.1, z=0.3))
```
Please add rotation to the base arm pose. The base arm pose seems to be rotated 180 degrees in the x-axis.

MagdalenaKotynia commented on src/rai_bench/rai_bench/o3de_test_bench/tasks/__init__.py:
```diff
@@ -0,0 +1,17 @@
+from rai_bench.o3de_test_bench.tasks.grab_carrot_task import GrabCarrotTask
+from rai_bench.o3de_test_bench.tasks.place_cubes_task import PlaceCubesTask
+
+__all__ = ["GrabCarrotTask", "PlaceCubesTask"]
```
What is the purpose of using __ all __ here?

jmatejcz replied:
it is here so the imports of tasks are easier - oneline, instead of separate import for every task

MagdalenaKotynia commented on src/rai_bench/rai_bench/o3de_test_bench/tasks/grab_carrot_task.py:
```diff
@@ -0,0 +1,99 @@
+    def get_prompt(self) -> str:
+        return "Manipulate objects, so that all carrots to the left side of the table (positive y)"
+
+    def validate_scene(self, simulation_config: SimulationConfig) -> bool:
```
I don't catch the sense of this method. It only verifies if the "carrot" prefab is present in the SimulationConfig. This check does not guarantee that the prefab named "carrot" is present in the scene. This make the implementation of this method inconsistent with the docstring of abstract method. 
Aprat from that, IMO checking if the specific prefab is available, should be handled by the setup_scene method of the SimulationBridge. On the example of O3DExROS2Bridge, if the prefab named "carrot" will not be available for specific scene, the _spawn_entity method will throw [error](https://github.com/RobotecAI/rai/blob/929ca5c1832f1460f0d602031ffadcf2bf5563e6/src/rai_sim/rai_sim/o3de/o3de_bridge.py#L158).

Therefore, I suggest removing the validate_scene method.

MagdalenaKotynia replied:
Update after private discussion: change name and docstring to reflect functionality of the method.

jmatejcz replied:
renamed function and removed example from docstring in abstract Task class

MagdalenaKotynia commented on src/rai_bench/rai_bench/benchmark_model.py:
```diff
@@ -0,0 +1,246 @@
+        return adjacent_count
+
+
+class Scenario:
```
What was the purpose of converting the BaseModel class to the standard class? 

Apart from my suggestion to remove the validate_scene method in the previous comment, such validation could be done in the BaseModel class (see Pydantic [validators](https://docs.pydantic.dev/2.0/usage/validators/))

jmatejcz replied:
the purpouse was to validate in constructor, apart from that it makes more sense to me that Scenario is normal class, not a pydatnic model

jmatejcz committed:
feat: added verification if required services , toipcs and actions are available

MagdalenaKotynia commented on src/rai_bench/rai_bench/benchmark_model.py:
```diff
@@ -0,0 +1,246 @@
+                    "initial_score": initial_result,
+                    "final_score": result,
+                    "total_time": f"{total_time:.3f}",
+                    "numer_of_tool_calls": tool_calls_num,
```
```suggestion
                    "number_of_tool_calls": tool_calls_num,
```

jmatejcz replied:
applied

jmatejcz commented:
> the scenario should not run until all services are up ( case when grounding dino is downloading weights)

propsed solution in https://github.com/RobotecAI/rai/pull/426/commits/ac14185b84811e57850665aaa8938540857ac68b

MagdalenaKotynia commented on .gitignore:
```diff
@@ -172,3 +172,6 @@ logs/
 artifact_database.pkl
+
+imgui.ini
+o3de_config.yaml
```
This file is not included in the repo (what is good) and it is user's specific config file, so I think that it should not be added to .gitignore.

jmatejcz replied:
removed config from gitignore

MagdalenaKotynia reviewed (changes requested):
I left some comments. Apart from that, there is quite a lot of missed/wrong typing. Could you please fix it and consider switching to a more strict typing checking mode?

jmatejcz committed:
refactor: funtions naming, typing and rotation of arm change

jmatejcz committed:
style: change of formatting

jmatejcz committed:
docs: extended readme

MagdalenaKotynia commented on src/rai_bench/rai_bench/benchmark_model.py:
```diff
@@ -0,0 +1,249 @@
+    SpawnedEntity,
+)
+
+SimulationBridgeT = TypeVar(
```
I see that this is not used anywhere, please remove it.

jmatejcz replied:
deleted it

MagdalenaKotynia commented on src/rai_sim/rai_sim/o3de/o3de_bridge.py:
```diff
@@ -199,6 +205,35 @@ def get_scene_state(self) -> SceneState:
+    def _is_robotic_stack_ready(
+        self, simulation_config: O3DExROS2SimulationConfig
+    ) -> bool:
+        retries = 0
```
Please parametrize number of retires with some default value

jmatejcz replied:
done

MagdalenaKotynia commented on src/rai_sim/rai_sim/o3de/o3de_bridge.py:
```diff
@@ -199,6 +205,35 @@ def get_scene_state(self) -> SceneState:
+                self.logger.info("All required services are available.")
+                return True
+
+            time.sleep(5)
```
Is such big time sleep needed?

jmatejcz replied:
the download of the model and other setups can last a long time, no point in checking more frequently IMO

MagdalenaKotynia commented on src/rai_bench/rai_bench/benchmark_model.py:
```diff
@@ -0,0 +1,249 @@
+    @classmethod
+    def create_scenarios(
+        cls, tasks: List[Task], simulation_configs: List[SimulationConfig]
+    ) -> list[Any]:
```
Why is it typed to return list[Any]?

jmatejcz replied:
you already fixed it, thanks ;)

MagdalenaKotynia committed:
fix: fix of wrong generic classes typing

jmatejcz committed:
style: delete unsed genric

jmatejcz committed:
refactor: parametrized number of retries in checking robotic stack

jmatejcz committed:
style: add logs about how many scenarios left
added log when all scenarios finished

jmatejcz committed:
feat: add dumping results to file
refactored scenario to inlucde config path so it can be logged easily

rachwalk reviewed (approved):
LGTM
